conn/osxkeychain.c

```diff
@@ -0,0 +1,209 @@
+ * this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * @page conn_osxkeychain Read/write macOS Keychain credentials
+ *
+ * Called by mutt_account_getuser() and mutt_user_getpass() as an alternative to
+ * storing/retrieving credentials in the configutation file, or entering them in
+ * the prompt repeatedly. A dialog will automatically pop up requesting permissions
+ * on macOS the very first time.
+ *
+ * Enabled using the configure option '--osxkeychain'.
+ */
+
+#include <Security/Security.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "mutt/lib.h"
+#include "connaccount.h"
+#include "mutt_account.h"
+
+/**
+ * struct Credential - A bunch of pointers that act as a lens into ConnAccount,
+ *                     filtering out only the information relevant to keychain.
+ */
+struct Credential
+{
+  SecProtocolType protocol; ///< Connection type, e.g. HTTP
+  UInt16 port;              ///< Port to connect to
+  const char *host;         ///< Server to login to
+  const char *user;         ///< Username
+  const char *pass;         ///< Password
+};
+
+static struct Credential cred;
+
+#define KEYCHAIN_ITEM(x) (x ? strlen(x) : 0), x
+#define KEYCHAIN_ARGS                                                          \
+  NULL,                                  /* default keychain */                \
+      KEYCHAIN_ITEM(cred.host), 0, NULL, /* account domain */                  \
+      KEYCHAIN_ITEM(cred.user), 0, NULL, cred.port, cred.protocol, kSecAuthenticationTypeDefault
+
+/**
+ * find_username_in_item - Find the username in a given SecKeychainItem
+ * @param item A reference to the security keychain item
+ * @param user A pre-allocated buffer into which to write the username
+ * @retval 0 on success, a specialized error-code on failure
+ */
+static int find_username_in_item(SecKeychainItemRef item, char *user)
+{
+  SecKeychainAttributeList list = { 0 };
+  SecKeychainAttribute attr = { 0 };
+
+  list.count = 1;
+  list.attr = &attr;
+  attr.tag = kSecAccountItemAttr;
+
+  int ret = SecKeychainItemCopyContent(item, NULL, &list, NULL, NULL);
+  if (ret != 0)
+    return ret;
+
+  mutt_str_copy(user, attr.data, attr.length + 1);
+  SecKeychainItemFreeContent(&list, NULL);
+  return ret;
+}
+
+/**
+ * find_internet_password - Find a username/password pair in keychain
+ * @param user A pre-allocated buffer into which to write the username
+ * @param pass A pre-allocated buffer into which to write the password
+ * @retval 0 on success, a specialized error-code on failure
+ */
+static int find_internet_password(char *user, char *pass)
+{
+  void *buf = NULL;
+  UInt32 len = 0;
+  SecKeychainItemRef item = NULL;
+  int ret = 0;
+
+  ret = SecKeychainFindInternetPassword(KEYCHAIN_ARGS, &len, &buf, &item);
+  if (ret != 0)
+    return ret;
+
+  // Keychain does not send the password with null-terminator
+  mutt_str_copy(pass, buf, len + 1);
```

Comment:
OK, that's a bug!
A have a fixed-length buffer, `pass`, but we don't check if `len + 1` might overflow.